File: api-handler/src/metrics.ts
/**
 * Prometheus metrics for api-handler
 */
import client from 'prom-client';

// Create a Registry to register the metrics
export const register = new client.Registry();

// Add default metrics
client.collectDefaultMetrics({ register });

// Custom metrics
export const stockQuoteRequests = new client.Counter({
  name: 'stock_quote_requests_total',
  help: 'Total number of stock quote requests',
  labelNames: ['symbol', 'provider', 'status'], // provider: brapi/yahoo, status: success/failure
  registers: [register],
});

export const stockHistoryRequests = new client.Counter({
  name: 'stock_history_requests_total',
  help: 'Total number of stock history requests',
  labelNames: ['symbol', 'provider', 'status'],
  registers: [register],
});

export const providerResponseTime = new client.Histogram({
  name: 'provider_response_time_seconds',
  help: 'Response time from stock data providers',
  labelNames: ['provider'],
  buckets: [0.1, 0.5, 1, 2, 5, 10],
  registers: [register],
});

export const providerErrors = new client.Counter({
  name: 'provider_errors_total',
  help: 'Total number of errors from stock data providers',
  labelNames: ['provider', 'error_type'],
  registers: [register],
});

File: scraping-worker/src/metrics.py
"""
Prometheus metrics for scraping-worker
"""
from prometheus_client import Counter, Gauge, Histogram, CollectorRegistry

# Create a registry
registry = CollectorRegistry()

# Custom metrics
scraping_requests = Counter(
    'scraping_requests_total',
    'Total number of scraping requests',
    ['status'],  # success/failure
    registry=registry
)

dividends_found = Counter(
    'dividends_found_total',
    'Total number of dividends found during scraping',
    ['symbol'],
    registry=registry
)

fundamental_data_scraped = Counter(
    'fundamental_data_scraped_total',
    'Total number of fundamental data records scraped',
    ['symbol', 'status'],
    registry=registry
)

scraping_duration = Histogram(
    'scraping_duration_seconds',
    'Time taken to scrape a single symbol',
    ['symbol'],
    buckets=[1, 2, 5, 10, 30, 60],
    registry=registry
)

scraping_errors = Counter(
    'scraping_errors_total',
    'Total number of scraping errors',
    ['error_type'],
    registry=registry
)

status_invest_requests = Counter(
    'status_invest_requests_total',
    'Total requests made to StatusInvest',
    ['status'],
    registry=registry
)
